core/src/main/java/io/questdb/cairo/wal/TableNameRegistryRW.java

```diff
@@ -0,0 +1,118 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.wal;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.CairoException;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.Chars;
+import io.questdb.std.ConcurrentHashMap;
+
+public class TableNameRegistryRW extends AbstractTableNameRegistry {
+    private static final Log LOG = LogFactory.getLog(TableNameRegistryRW.class);
+    private final ConcurrentHashMap<String> reverseTableNameCache = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<TableNameRecord> systemTableNameCache = new ConcurrentHashMap<>();
+
+    public TableNameRegistryRW(CairoConfiguration configuration) {
+        super(configuration);
+        if (!this.tableNameStore.lock()) {
+            if (!configuration.getAllowTableRegistrySharedWrite()) {
+                throw CairoException.critical(0).put("cannot lock table name registry file [path=").put(configuration.getRoot()).put(']');
+            }
+        }
+        setNameMaps(systemTableNameCache, reverseTableNameCache);
+    }
+
+    @Override
+    public void deleteNonWalName(CharSequence tableName, String systemTableName) {
+        systemTableNameCache.remove(tableName);
+        reverseTableNameCache.remove(systemTableName);
+    }
+
+    @Override
+    public String registerName(String tableName, String systemTableName, boolean isWal) {
+        TableNameRecord newNameRecord = new TableNameRecord(systemTableName, isWal);
+        TableNameRecord registeredRecord = systemTableNameCache.putIfAbsent(tableName, newNameRecord);
+
+        if (registeredRecord == null) {
+            if (isWal) {
+                tableNameStore.appendEntry(tableName, systemTableName);
+            }
+            reverseTableNameCache.put(systemTableName, tableName);
+            return systemTableName;
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public synchronized void reloadTableNameCache() {
+        LOG.info().$("reloading table to system name mappings").$();
+        systemTableNameCache.clear();
+        reverseTableNameCache.clear();
+        tableNameStore.reload(systemTableNameCache, reverseTableNameCache, TABLE_DROPPED_MARKER);
+    }
+
+    @Override
+    public void removeTableSystemName(CharSequence systemTableName) {
+        reverseTableNameCache.remove(systemTableName);
+    }
+
+    @Override
+    public boolean removeWalTableName(CharSequence tableName, String systemTableName) {
```

Comment:
if this races with creating new table registry will be corrupt, it looks very unsafe to use multiple concurrent maps

Reply:
Added tests in `TableNameRegistryTest` which concurrently create and drop tables